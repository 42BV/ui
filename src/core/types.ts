/* istanbul ignore file */

export type Color =
  | 'primary'
  | 'secondary'
  | 'success'
  | 'info'
  | 'warning'
  | 'danger'
  | 'link'
  | 'muted'
  | 'dark'
  | 'light';

export type ModalSize = 'sm' | 'md' | 'lg' | 'xl';

type Side = 'top' | 'right' | 'bottom' | 'left';
type StartOrEnd = 'start' | 'end';
export type TippyPlacement = Side | `${Side}-${StartOrEnd}`;
