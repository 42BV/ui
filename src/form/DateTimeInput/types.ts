// See https://momentjs.com/docs/#/displaying/format/ for formatting options.
export type DateFormat = string | false; // When false it should not be used
export type TimeFormat = string | false; // When false it should not be used
