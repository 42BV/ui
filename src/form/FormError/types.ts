export type OnChange = (hasErrors: boolean) => void;