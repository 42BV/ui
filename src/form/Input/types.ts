export type Position = 'left' | 'right';
export type Mask = (string | RegExp)[];
