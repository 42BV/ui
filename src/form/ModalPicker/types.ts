export type AddButtonCallback<T> = () => Promise<T>;

export interface AddButtonOptions<T> {
  label: string;
  onClick: AddButtonCallback<T>;
}
