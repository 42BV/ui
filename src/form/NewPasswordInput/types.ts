export type NewPasswordInputRule =
  | 'lowercase'
  | 'uppercase'
  | 'number'
  | 'specialChar'
  | 'minimumLength'
  | 'noSpace';
