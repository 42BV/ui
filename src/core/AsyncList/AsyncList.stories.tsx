import React from 'react';
import { storiesOf } from '@storybook/react';
import { useAsync } from 'react-async';
import { Button, Card, ListGroup, ListGroupItem } from 'reactstrap';

import AsyncList from './AsyncList';

import { pageOfUsers } from '../../test/fixtures';
import { User } from '../../test/types';
import { ContentState } from '../..';
import { action } from '@storybook/addon-actions';

function loadData(): Promise<User[]> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve(pageOfUsers.content);
    }, 1000);
  });
}

function emptyData(): Promise<User[]> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve([]);
    }, 1000);
  });
}

function rejectData(): Promise<User[]> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      reject('could not load');
    }, 1000);
  });
}

function loadingData(): Promise<User[]> {
  return new Promise(() => undefined);
}

storiesOf('core|Async/AsyncList', module)
  .addParameters({ component: AsyncList })
  .add('when loaded', () => {
    const state = useAsync(loadData);

    return (
      <Card body>
        <AsyncList state={state}>
          {users => (
            <ListGroup>
              {users.map(user => (
                <ListGroupItem key={user.id}>{user.email}</ListGroupItem>
              ))}
            </ListGroup>
          )}
        </AsyncList>
      </Card>
    );
  })

  .add('when error', () => {
    const state = useAsync(rejectData);

    return (
      <Card body>
        <AsyncList state={state}>
          {users => (
            <ListGroup>
              {users.map(user => (
                <ListGroupItem key={user.id}>{user.email}</ListGroupItem>
              ))}
            </ListGroup>
          )}
        </AsyncList>
      </Card>
    );
  })

  .add('when error with custom text', () => {
    const state = useAsync(rejectData);

    return (
      <Card body>
        <AsyncList
          state={state}
          text={{ error: 'I’m sorry Dave, I’m afraid I can’t do that' }}
        >
          {users => (
            <ListGroup>
              {users.map(user => (
                <ListGroupItem key={user.id}>{user.email}</ListGroupItem>
              ))}
            </ListGroup>
          )}
        </AsyncList>
      </Card>
    );
  })

  .add('when error with no retry button', () => {
    const state = useAsync(rejectData);

    return (
      <Card body>
        <AsyncList state={state} showRetryButton={false}>
          {users => (
            <ListGroup>
              {users.map(user => (
                <ListGroupItem key={user.id}>{user.email}</ListGroupItem>
              ))}
            </ListGroup>
          )}
        </AsyncList>
      </Card>
    );
  })

  .add('when loading', () => {
    const state = useAsync(loadingData);

    return (
      <Card body>
        <AsyncList state={state}>
          {users => (
            <ListGroup>
              {users.map(user => (
                <ListGroupItem key={user.id}>{user.email}</ListGroupItem>
              ))}
            </ListGroup>
          )}
        </AsyncList>
      </Card>
    );
  })

  .add('when loading with custom title', () => {
    const state = useAsync(loadingData);

    return (
      <Card body>
        <AsyncList state={state} text={{ loading: 'Loading Jeffrey' }}>
          {users => (
            <ListGroup>
              {users.map(user => (
                <ListGroupItem key={user.id}>{user.email}</ListGroupItem>
              ))}
            </ListGroup>
          )}
        </AsyncList>
      </Card>
    );
  })

  .add('when empty', () => {
    const state = useAsync(emptyData);

    return (
      <Card body>
        <AsyncList state={state}>
          {users => (
            <ListGroup>
              {users.map(user => (
                <ListGroupItem key={user.id}>{user.email}</ListGroupItem>
              ))}
            </ListGroup>
          )}
        </AsyncList>
      </Card>
    );
  })

  .add('when empty with custom title', () => {
    const state = useAsync(emptyData);

    return (
      <Card body>
        <AsyncList
          state={state}
          text={{ empty: "No Jeffrey's match your parameters try again" }}
        >
          {users => (
            <ListGroup>
              {users.map(user => (
                <ListGroupItem key={user.id}>{user.email}</ListGroupItem>
              ))}
            </ListGroup>
          )}
        </AsyncList>
      </Card>
    );
  })

  .add('when empty with custom empty', () => {
    const state = useAsync(emptyData);

    return (
      <Card body>
        <AsyncList
          state={state}
          emptyContent={() => (
            <ContentState mode="empty" title="No results found">
              <Button icon="refresh" onClick={action('clear filters')}>
                Clear filters
              </Button>
            </ContentState>
          )}
        >
          {users => (
            <ListGroup>
              {users.map(user => (
                <ListGroupItem key={user.id}>{user.email}</ListGroupItem>
              ))}
            </ListGroup>
          )}
        </AsyncList>
      </Card>
    );
  });
