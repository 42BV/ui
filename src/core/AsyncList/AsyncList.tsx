import React from 'react';

import AsyncContent, {
  Props as AsyncContentProps
} from '../AsyncContent/AsyncContent';

type Props<T> = Omit<AsyncContentProps<T>, 'isEmpty'>;

export function isEmpty<T>(list: T[]) {
  return list.length === 0;
}

/**
 * The `AsyncList` is a variant of the `AsyncContent` which expects
 * the result of a call to `useAsync` from `react-async` to be an
 * array.
 *
 * It has all the behaviors of the `AsyncContent` but provides a
 * standard `isEmpty` function which checks if the array has the
 * `length` zero.
 */
export default function AsyncList<T>(props: Props<T[]>) {
  return <AsyncContent {...props} isEmpty={isEmpty} />;
}
