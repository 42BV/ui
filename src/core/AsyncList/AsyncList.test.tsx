import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import AsyncList, { isEmpty } from './AsyncList';

import { User } from '../../../test/types';
import { pageOfUsers } from '../../../test/fixtures';

type State = {
  isLoading: boolean;
  isFulfilled: boolean;
  data: User[];
};

describe('Component: Async|AsyncList', () => {
  function setup({ state }: { state: State }) {
    const asyncList = shallow(
      <AsyncList
        // @ts-ignore
        state={state}
        showRetryButton={true}
        text={{ loading: 'Custom loading' }}
      >
        {users => (
          <ul>
            {users.map(user => (
              <li key={user.id}>{user.email}</li>
            ))}
          </ul>
        )}
      </AsyncList>
    );

    return { asyncList };
  }

  test('ui', () => {
    const state = {
      isLoading: false,
      isFulfilled: true,
      data: pageOfUsers.content
    };

    const { asyncList } = setup({ state });

    expect(toJson(asyncList)).toMatchSnapshot();
  });
});

describe('isEmpty', () => {
  it('should consider it empty when the totalElements is zero', () => {
    expect(isEmpty([])).toBe(true);
  });

  it('should consider it empty when the totalElements is zero', () => {
    expect(isEmpty(pageOfUsers.content)).toBe(false);
  });
});
