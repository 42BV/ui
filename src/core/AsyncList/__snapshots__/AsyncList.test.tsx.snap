// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Async|AsyncList ui 1`] = `
<div>
  <ul>
    <li>
      admin@42.nl
    </li>
    <li>
      coordinator@42.nl
    </li>
    <li>
      user@42.nl
    </li>
  </ul>
</div>
`;
